File: frontend/bff/src/infrastructure/grpc/mappers/order.contract-mapper.ts
import type { OrderTermsInput } from '../../../application/orders/commands/order-terms.input.js';
import type { OrderDto } from '../../../application/orders/dtos/order.dto.js';
import { nullable, nullableNumber } from './common.contract-mapper.js';
import {
  approvalStatusEnum,
  commodityEnum,
  complianceStatusEnum,
  confirmationStatusEnum,
  optionKindEnum,
  orderTypeEnum,
  tradeDirectionEnum,
} from './enum.contract-mapper.js';

export interface ContractOrderTerms {
  type: string;
  direction: string;
  tenor: string;
  lots?: number;
  notionalUsd?: number;
  price: number;
  priceUnit?: string;
  optionKind: string;
  premium?: number;
  tradeDate: string;
  notes?: string;
  commodity?: string;
  coveredSale?: boolean;
  justification?: string;
}

export interface ContractOrder {
  id: string;
  code: string;
  mandateId?: string;
  mandateCode?: string;
  mandateTitle: string;
  counterpartyId: string;
  counterpartyName: string;
  commodity: string;
  terms: ContractOrderTerms;
  approval: string;
  requestedBy: string;
  decidedBy?: string;
  decidedAt?: string;
  decisionNote?: string;
  confirmation: string;
  confirmedOn?: string;
  confirmationNote?: string;
  confirmationOverdue: boolean;
  confirmationBy?: string;
  compliance: { status: string; reason: string } | null;
  linkedAfterExecution: boolean;
  exceedsMandate: boolean;
  deviationNote?: string;
}

export const toOrderDto = (o: ContractOrder): OrderDto => ({
  id: o.id,
  code: o.code,
  mandateId: nullable(o.mandateId),
  mandateCode: nullable(o.mandateCode),
  mandateTitle: o.mandateTitle,
  counterpartyId: o.counterpartyId,
  counterpartyName: o.counterpartyName,
  commodity: commodityEnum.fromContract(o.commodity),
  terms: {
    type: orderTypeEnum.fromContractRequired(o.terms.type),
    direction: tradeDirectionEnum.fromContractRequired(o.terms.direction),
    tenor: o.terms.tenor,
    lots: nullableNumber(o.terms.lots),
    notionalUsd: nullableNumber(o.terms.notionalUsd),
    price: o.terms.price,
    priceUnit: nullable(o.terms.priceUnit),
    optionKind: optionKindEnum.fromContract(o.terms.optionKind),
    premium: nullableNumber(o.terms.premium),
    tradeDate: o.terms.tradeDate,
    notes: nullable(o.terms.notes),
    coveredSale: o.terms.coveredSale ?? false,
  },
  approval: approvalStatusEnum.fromContractRequired(o.approval),
  requestedBy: o.requestedBy,
  decidedBy: nullable(o.decidedBy),
  decidedAt: nullable(o.decidedAt),
  decisionNote: nullable(o.decisionNote),
  confirmation: confirmationStatusEnum.fromContractRequired(o.confirmation),
  confirmedOn: nullable(o.confirmedOn),
  confirmationNote: nullable(o.confirmationNote),
  confirmationOverdue: o.confirmationOverdue,
  confirmationBy: nullable(o.confirmationBy),
  compliance: {
    status: complianceStatusEnum.fromContract(o.compliance?.status ?? '') === 'Outside' ? 'Outside' : 'Within',
    reason: o.compliance?.reason ?? '',
  },
  linkedAfterExecution: o.linkedAfterExecution,
  exceedsMandate: o.exceedsMandate,
  deviationNote: nullable(o.deviationNote),
});

export const toContractOrderTerms = (t: OrderTermsInput) => ({
  ...t,
  type: orderTypeEnum.toContract(t.type),
  direction: tradeDirectionEnum.toContract(t.direction),
  optionKind: optionKindEnum.toContract(t.optionKind),
  commodity: commodityEnum.toContract(t.commodity),
});

File: backend/core-service/src/Fix.Presentation/Mappers/OrderContractMapper.cs
using Fix.Application.Orders.Commands;
using Fix.Application.Orders.Dtos;
using Contract = Fix.Contracts.V1;
using Entities = Fix.Domain.AggregateRoots.Orders;

namespace Fix.Presentation.Mappers;

internal static class OrderContractMapper
{
    public static Contract.Order ToContract(this OrderDto order)
    {
        var terms = new Contract.OrderTerms
        {
            Type = order.Type.ToContract<Contract.OrderType>(),
            Direction = order.Direction.ToContract<Contract.TradeDirection>(),
            Tenor = order.Tenor,
            Price = (double)order.Price,
            PriceUnit = order.PriceUnit,
            OptionKind = order.OptionKind.ToContract<Contract.OptionKind>(),
            TradeDate = order.TradeDate.ToContract(),
        };

        if (order.Lots is { } lots) terms.Lots = (double)lots;
        if (order.NotionalUsd is { } notional) terms.NotionalUsd = (double)notional;
        if (order.Premium is { } premium) terms.Premium = (double)premium;
        if (order.Notes is { } notes) terms.Notes = notes;

        var contract = new Contract.Order
        {
            Id = order.Id.ToString(),
            MandateId = order.MandateId.ToString(),
            MandateTitle = order.MandateTitle,
            CounterpartyId = order.CounterpartyId.ToString(),
            CounterpartyName = order.CounterpartyName,
            Commodity = order.Commodity.ToContract<Contract.Commodity>(),
            Terms = terms,
            Approval = order.Approval.ToContract<Contract.ApprovalStatus>(),
            RequestedBy = order.RequestedBy.ToString(),
            Confirmation = order.Confirmation.ToContract<Contract.ConfirmationStatus>(),
            ConfirmationOverdue = order.ConfirmationOverdue,
        };

        if (order.DecidedBy is { } decidedBy) contract.DecidedBy = decidedBy.ToString();
        if (order.DecidedAt is { } decidedAt) contract.DecidedAt = decidedAt.ToContract();
        if (order.DecisionNote is { } note) contract.DecisionNote = note;
        if (order.ConfirmedOn is { } confirmedOn) contract.ConfirmedOn = confirmedOn.ToContract();
        if (order.ConfirmationNote is { } confirmationNote) contract.ConfirmationNote = confirmationNote;
        return contract;
    }

    public static OrderTermsInput ToInput(this Contract.OrderTerms? terms)
    {
        terms ??= new Contract.OrderTerms();
        return new OrderTermsInput(
            terms.Type.ToDomain<Entities.OrderType>("terms.type"),
            terms.Direction.ToDomain<Entities.TradeDirection>("terms.direction"),
            terms.Tenor,
            terms.Lots.ToOptionalDecimal(terms.HasLots),
            terms.NotionalUsd.ToOptionalDecimal(terms.HasNotionalUsd),
            terms.Price.ToDecimal(),
            terms.PriceUnit.ToOptionalString(terms.HasPriceUnit),
            terms.OptionKind.ToOptionalDomain<Entities.OptionKind>("terms.option_kind"),
            terms.Premium.ToOptionalDecimal(terms.HasPremium),
            terms.TradeDate.ToDate("terms.trade_date"),
            terms.Notes.ToOptionalString(terms.HasNotes));
    }
}

